### PR TITLE
Refactor `tabular_predictions.py`

### PR DESCRIPTION
**Pull request by Innixma**
Major refactor of `tabular_predictions.py`

Highlights:

1. Added various restriction methods and force_to_dense methods to ensure validity of the data prior to ZS simulation.
2. Added optimized variant of the predictions format which is far faster in simulation with Ray due to having fewer overall numpy arrays. (~2x-3x speedup in ZS-sim speed, identical results).
3. Standardized the capabilities of all prediction class variants.
4. Added unit tests for various advanced logic to guard against future bugs.
5. Various code cleanup / bug fixes

Note: These changes build upon #12, but are significant enough that I figured it would be better as its own PR.

### CODE REVIEW

File: autogluon_zeroshot/contexts/context_2023_03_19_bag_289.py [hunk 1]
```diff
@@ -31,14 +32,18 @@ def load_context_2023_03_19_bag_289(
...
-        results_path = str(path_bagged_root / "608/results_ranked_valid.parquet")
-        results_by_dataset_path = str(path_bagged_root / "608/results_ranked_by_dataset_valid.parquet")
-        raw_path = str(path_bagged_root / "openml_ag_2023_03_19_zs_models.parquet")
-    else:
+    results_path = str(path_bagged_root / "608/results_ranked_valid.parquet")
+    results_by_dataset_path = str(path_bagged_root / "608/results_ranked_by_dataset_valid.parquet")
+    raw_path = str(path_bagged_root / "openml_ag_2023_03_19_zs_models.parquet")
+    path_automl = str(Paths.automl_289_results_root / "results_ranked_by_dataset_valid.csv")
+    local_files_exist = [Path(f).exists() for f in [results_path, results_by_dataset_path, raw_path, path_automl]]
+    if False in local_files_exist:
```
Comment (geoalgo) [inline autogluon_zeroshot/contexts/context_2023_03_19_bag_289.py]: Good catch, we could use a more pythonic way:
```suggestion
    local_files_exist = all(Path(f).exists() for f in [results_path, results_by_dataset_path, raw_path, path_automl])
    if not local_files_exist:
```

File: autogluon_zeroshot/contexts/utils.py [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
-    zeroshot_pred_proba.force_to_dense(prune_method='dataset')
+    zeroshot_pred_proba.force_to_dense(first_prune_method='task', second_prune_method='dataset')
 
     zsc.subset_models(zeroshot_pred_proba.models)
     zsc.subset_datasets(zeroshot_pred_proba.datasets)
     zeroshot_pred_proba.restrict_models(zsc.get_configs())
     zeroshot_gt = prune_zeroshot_gt(zeroshot_pred_proba=zeroshot_pred_proba, zeroshot_gt=zeroshot_gt)
 
+    # Convert to optimized format
+    if isinstance(zeroshot_pred_proba, TabularPicklePredictions) and \
```
Comment (geoalgo) [inline autogluon_zeroshot/contexts/utils.py:26]: Do we need the type test? I did not see how the type could be `TabularPicklePredictionsOpt` at this point.

File: autogluon_zeroshot/simulation/simulation_context.py [hunk 3]
```diff
@@ -213,7 +215,7 @@ def load_pred(self, pred_pkl_path: Path, lazy_format: bool = False) -> dict:
 
     @staticmethod
     def convert_lazy_format(pred_pkl_path, override_if_already_exists: bool = False):
-        new_filename = pred_pkl_path.parent / "zeroshot_pred_per_task"
+        new_filename = Path(str(pred_pkl_path).rsplit('.pkl', 1)[0])
```
Comment (geoalgo) [inline autogluon_zeroshot/simulation/simulation_context.py]: Why is this change needed? If it is needed, can we use Path rather to avoid potential portability issue? (eg `Path(pred_pkl_path).parent / Path(pred_pkl_path).stem`)

File: autogluon_zeroshot/simulation/tabular_predictions.py [hunk 4]
```diff
@@ -35,22 +45,229 @@ def predict(self, dataset: str, fold: int, splits: List[str] = None, models: Lis
...
+              f'\tis_dense={is_dense} | is_dense_folds={is_dense_folds} | is_dense_models={is_dense_models}')
 
     @property
     def models(self) -> List[str]:
+        """
+        :return: list of models present in at least one dataset.
+        """
+        raise NotImplementedError()
+
+    def get_models(self, present_in_all=False) -> List[str]:
```
Comment (geoalgo) [inline autogluon_zeroshot/simulation/tabular_predictions.py:81]: As discussed offline, lets merge it as the changes are becoming quite significant. I will work on separating most of those methods as free functions into a densifier util.

File: autogluon_zeroshot/simulation/tabular_predictions.py [hunk 5]
```diff
@@ -362,25 +710,24 @@ def rename_datasets(self, rename_dict: dict):
...
     def _save_metadata(output_dir, dataset_to_models):
-        with open(output_dir / "metadata.json", "w") as f:
-            metadata = {
-                "dataset_to_models": dataset_to_models,
-            }
-            json.dump(metadata, f)
+        metadata = {
+            "dataset_to_models": dataset_to_models,
+        }
+        save_pkl(path=str(output_dir / "metadata.pkl"), object=metadata)
```
Comment (geoalgo) [inline autogluon_zeroshot/simulation/tabular_predictions.py]: If there is no unit test or issue that we can pinpoint to, can we keep json? Pkl is not safe without precaution as opposed to json (also json is readable in clear).

File: autogluon_zeroshot/simulation/tabular_predictions.py [hunk 6]
```diff
@@ -1,17 +1,27 @@
...
 import numpy as np
 from pathlib import Path
 
 from autogluon.common.loaders import load_pkl
 from autogluon.common.savers.save_pkl import save as save_pkl
 
+from .task_predictions import ConfigPredictionsDict, TaskModelPredictions, TaskModelPredictionsOpt
+
+# dictionary mapping a particular fold of a dataset (a task) to split to config name to predictions
+TaskPredictionsDict = Dict[str, ConfigPredictionsDict]
```
Comment (geoalgo) [inline autogluon_zeroshot/simulation/tabular_predictions.py:18]: Very nice to add those types.

**Review by geoalgo — APPROVED** [review]
Feel free to merge after addressing the changes if its easier for you. I will send a PR after.

File: autogluon_zeroshot/simulation/simulation_context.py (hunk 3 above)
Comment (Innixma) [reply autogluon_zeroshot/simulation/simulation_context.py]: I sometimes have suffixes to the names of the pkl files, so this allows flexibility to simply have the folder be the name of the file rather than a fixed `zeroshot_pred_per_task`. Otherwise they would overwrite eachother.

```
    "results/bagged_289/zeroshot_pred_proba_10_mb.pkl": "s3://automl-benchmark-ag/aggregated/ec2/2023_03_19_zs/zeroshot_pred_proba_10_mb.pkl",
    "results/bagged_289/zeroshot_pred_proba_50_mb.pkl": "s3://automl-benchmark-ag/aggregated/ec2/2023_03_19_zs/zeroshot_pred_proba_50_mb.pkl",
```

+1 to using Path. This is generally something I should consider using more often rather than str for these path operations, I wasn't very familiar with them prior to you introducing me to it.

Comment (Innixma) [reply autogluon_zeroshot/simulation/simulation_context.py]: Updated to use your recommended code snippet

File: autogluon_zeroshot/contexts/utils.py (hunk 2 above)
Comment (Innixma) [reply autogluon_zeroshot/contexts/utils.py:26]: Technical not, but it is possible `zsc.load_pred` could in future return `TabularPicklePredictionsOpt`. At minimum the type check doesn't hurt.

File: autogluon_zeroshot/simulation/tabular_predictions.py (hunk 5 above)
Comment (Innixma) [reply autogluon_zeroshot/simulation/tabular_predictions.py]: This is the problem: https://stackoverflow.com/questions/1450957/pythons-json-module-converts-int-dictionary-keys-to-strings

Comment (Innixma) [reply autogluon_zeroshot/simulation/tabular_predictions.py]: Example of how this breaks our current code:

```python3
import json

dataset_to_models = {
    377590: {  # dataset TID
        0: ['model1', 'model2'],  # fold to models
        1: ['model1', 'model2'],  # fold to models
    }
}
json_dataset_to_models = json.dumps(dataset_to_models)

load_dataset_to_models = json.loads(json_dataset_to_models)

print(f'Original:')
print(dataset_to_models)
print(f'JSON Loaded:')
print(load_dataset_to_models)

assert dataset_to_models == load_dataset_to_models

```

Out:

```
Original:
{377590: {0: ['model1', 'model2'], 1: ['model1', 'model2']}}
JSON Loaded:
{'377590': {'0': ['model1', 'model2'], '1': ['model1', 'model2']}}
Traceback (most recent call last):
  File "/home/ubuntu/workspace/code/scratch/scratch1.py", line 18, in <module>
    assert dataset_to_models == load_dataset_to_models
AssertionError
```

Comment (Innixma) [reply autogluon_zeroshot/simulation/tabular_predictions.py]: If you prefer to keep using JSON, we would need to either:

1. Figure out how to do it with integers kept in tact (seems not possible without a lot of hacking)
2. Convert all integer keys to strings (would prefer not doing this though)
